gdal/gcore/gdal_priv.h

```diff
@@ -659,20 +660,32 @@ class CPL_DLL GDALDataset : public GDALMajorObject
                 struct Private;
                 std::unique_ptr<Private> m_poPrivate;
             public:
+
+                using value_type = OGRLayer*;
+                using reference = OGRLayer*;
+                using difference_type = void;
+                using pointer = void;
+                using iterator_category = std::input_iterator_tag;
+
+                Iterator();
```

Comment:
given that this Iterator class is a private class of GDALDataset, there's noway that the Iterator() constructor can be called currently.

Reply:
Ah, I overlooked that. Why would you want the Iterator class to be private? It is part of the public interface as the return type of begin() and end(). 

So actually, with some trickery you can call the constructor:

```cpp
// deducing the type
using my_iterator_type = decltype(std::declval<GDALDataset::Layers>()::begin());

// applying the constructor
my_iterator_type  i;
```

See also this standalone example:
```cpp
#include <iostream>
class A {
  class B 
  {
  public:
    B() { std::cout << "Default constructor" << std::endl;  }
  };

public:
  B operator()() { return B{}; }
};

int main()
{
  A a;
  using b_type = decltype(a());
  b_type b;
}
```

Reply:
Good point. The idea was that if you do a range-based loop, you never manipulate the iterator explicitly, so it can be relatively hidden. But might make sense to make it public if it is now a 'real' iterator. While require some Doxygen documentation to please the CI tests;